### ISSUE TITLE
Avoid capturing unnecessary constrained variables in `FunctionClosure` phase

### ISSUE THREAD

**Issue by mario-bucev** [epfl-lara/stainless#1259]
The following `inner` function unnecessarily captures `x` when lifted:
```scala
object Test {
  def outer(x: Int): Unit = {
    require(x >= 0)
    def inner(y: Int): Unit = ()
  }
}
```
is roughly transformed to:
```scala
def outer(x: Int): Unit = {
  require(x >= 0)
}

def inner(x: Int, y: Int): Unit = {
  require(x >= 0)
}
```
(due to `x` being constrained with the `require`).

A naive approach checking for occurrences is not enough though, in the following example, we should not only capture `a` but also `x` (due to constraining `a`): 
```scala
object Test {
  def outer(x: Int, a: Int): Unit = {
    require(x >= 0 && a >= x)
    def inner(y: Int): Unit = {
      val b = a // Captures a, but the hoisting of inner must also capture x for the PC a >= x (and x >= 0 as well)
    }
  }
}
```

**Comment by samarion** [comment]
We might drop cases where we have `require(false)` with this change. This could be relevant for potential transformations which _e.g._ introduce case reasoning. I don't think we actually have any currently, just something to think about.

What is your use case for such a change?

**Comment by mario-bucev** [comment]
Sorry, I forgot to justify!
In the last Friday's meeting, we discussed the equivalence checker, and Dragana pointed out that it may fail on some programs containing simple `while` loops. She reduced the issue down to the hoisting of the transformed `while`-functions unnecessarily capturing some variables.
If two `while`-functions are equivalent but happen to capture different, unrelated variables, the EC rejects the program due to these functions not having the same signature.
We have discussed possible solutions, and have come to open this ticket :)

(I am unsure about whether the EC actually rejects or timeouts, but it does not output an expected result, btw @drganam feel free to correct me if I'm wrong on something)
